onvif/src/discovery/mod.rs

```diff
@@ -43,121 +46,163 @@ impl Debug for Device {
-/// # Examples
-///
-/// You can access each element on the stream concurrently as soon as devices respond:
-///
-/// ```
-/// use onvif::discovery;
-/// use futures_util::stream::StreamExt; // to use for_each_concurrent
-///
-/// const MAX_CONCURRENT_JUMPERS: usize = 100;
-///
-/// async {
-///     discovery::discover(std::time::Duration::from_secs(1))
-///         .await
-///         .unwrap()
-///         .for_each_concurrent(MAX_CONCURRENT_JUMPERS, |addr| {
-///             async move {
-///                 println!("Device found: {:?}", addr);
-///             }
-///         })
-///         .await;
-/// };
-/// ```
-///
-/// Or you can await on a collection of unique devices found in one second:
-///
-/// ```
-/// use onvif::discovery;
-/// use futures_util::stream::StreamExt; // to use collect
-/// use std::collections::HashSet;
-///
-/// async {
-///     let devices = discovery::discover(std::time::Duration::from_secs(1))
-///         .await
-///         .unwrap()
-///         .collect::<HashSet<_>>()
-///         .await;
-///
-///     println!("Devices found: {:?}", devices);
-/// };
-/// ```
-pub async fn discover(duration: Duration) -> Result<impl Stream<Item = Device>, Error> {
-    let probe = Arc::new(build_probe());
-    let probe_xml = yaserde::ser::to_string(probe.as_ref()).map_err(Error::Serde)?;
-
-    debug!("Probe XML: {}", probe_xml);
-
-    let socket = {
-        const LOCAL_IPV4_ADDR: Ipv4Addr = Ipv4Addr::UNSPECIFIED;
-        const LOCAL_PORT: u16 = 0;
-
-        const MULTI_IPV4_ADDR: Ipv4Addr = Ipv4Addr::new(239, 255, 255, 250);
-        const MULTI_PORT: u16 = 3702;
-
-        let local_socket_addr = SocketAddr::new(IpAddr::V4(LOCAL_IPV4_ADDR), LOCAL_PORT);
-        let multi_socket_addr = SocketAddr::new(IpAddr::V4(MULTI_IPV4_ADDR), MULTI_PORT);
-
-        let socket = UdpSocket::bind(local_socket_addr).await?;
-        socket.join_multicast_v4(MULTI_IPV4_ADDR, LOCAL_IPV4_ADDR)?;
-        socket
-            .send_to(probe_xml.as_bytes(), multi_socket_addr)
-            .await?;
-
-        socket
-    };
-
-    let (device_sender, device_receiver) = channel(32);
-    let device_receiver = ReceiverStream::new(device_receiver);
-
-    let mut known_responses = HashSet::new();
-
-    let produce_devices = async move {
-        while let Ok((xml, src)) = recv_string(&socket).await {
-            if !known_responses.insert(calculate_hash(&xml)) {
-                debug!("Duplicate response from {src}, skipping ...");
-                continue;
+#[derive(Debug, Clone)]
+pub struct DiscoveryBuilder {
+    duration: Duration,
+    listen_address: IpAddr,
+}
+
+impl Default for DiscoveryBuilder {
+    fn default() -> Self {
+        Self {
+            duration: Duration::from_secs(5),
+            listen_address: IpAddr::V4(Ipv4Addr::UNSPECIFIED),
+        }
+    }
+}
+
+impl DiscoveryBuilder {
+    /// How long to listen for the responses from the network.
+    pub fn duration(&mut self, duration: Duration) -> &mut Self {
```

Comment:
Should it be called timeout?

Reply:
IMHO it's not a timeout conceptually because discovery always takes exactly this amount of time. It doesn't stop earlier in case of success, as there's no definition of success, we just don't know how many cameras are on the network.